release-23.2: mixedversion: run skip-version upgrades when supported

DarrylWong opened the pull request:
Backport 1/1 commits from #125869.

/cc @cockroachdb/release

---

This changes the `mixedversion` framework to randomly perform "skip-version" upgrades when supported. Specifically, v23.1 onwards supports skip-version upgrades, where a single rolling upgrade can be performed to upgrade Cockroach two major releases at a time.

This behaviour is still experimental, and it is necessary to set an environment variable to enable this feature.

Tests that perform logic specific to a certain upgrade path can disable this functionality by using the `DisableSkipVersionUpgrades` option or gating certain steps behind a `helper.IsSkipVersionUpgrade` check.

Fixes: #121957
Epic: none
Release note: None

Release Justification: test only change

Commit:
mixedversion: run skip-version upgrades when supported

This changes the `mixedversion` framework to randomly perform
"skip-version" upgrades when supported. minSupportedSkipVersionUpgrade
indicates the minimum version after which skip version upgrades
are supported. The diff between the current version and the
minSupportedSkipVersionUpgrade denotes the number of possible
version skips.

Specifically, v24.1 onwards supports skip-version upgrades,
where a single rolling upgrade can be performed to upgrade
Cockroach two major releases at a time.

Tests that perform logic specific to a certain upgrade path can
disable this functionality by using the `DisableSkipVersionUpgrades`
option or gating certain steps behind a `helper.IsSkipVersionUpgrade`
check.

Fixes: #121957
Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127167)
<!-- Reviewable:end -->

DarrylWong committed:
mixedversion: pass in cluster settings when restarting binary

This was erroneously deleted in #125869.

DarrylWong committed:
mixedversion: bump datadriven test build version

Bump the build version to 24.3 for TestTestPlanner. This is
needed for the skip_version test to properly skip versions.
